crates/assists/src/utils/insert_use.rs

```diff
@@ -63,27 +63,46 @@ impl ImportScope {
         }
     }
 
-    fn insert_pos_after_inner_attribute(&self) -> (InsertPosition<SyntaxElement>, AddBlankLine) {
-        // check if the scope has inner attributes, we dont want to insert in front of them
-        match self
-            .as_syntax_node()
-            .children()
-            // no flat_map here cause we want to short circuit the iterator
-            .map(ast::Attr::cast)
-            .take_while(|attr| {
-                attr.as_ref().map(|attr| attr.kind() == ast::AttrKind::Inner).unwrap_or(false)
-            })
-            .last()
-            .flatten()
-        {
-            Some(attr) => {
-                (InsertPosition::After(attr.syntax().clone().into()), AddBlankLine::BeforeTwice)
+    fn insert_pos_after_inner_elements(&self) -> (InsertPosition<SyntaxElement>, AddBlankLine) {
+        let mut last_inner_element = None;
+
+        for maybe_inner_element in self.as_syntax_node().children_with_tokens() {
+            match maybe_inner_element {
+                NodeOrToken::Node(maybe_inner_node) => {
+                    if is_inner_node(maybe_inner_node.clone()) {
+                        last_inner_element = Some(NodeOrToken::Node(maybe_inner_node))
+                    } else {
+                        if let Some(maybe_inner_token) = maybe_inner_node.first_token() {
```

Comment:
This check exists due to the fact that for
```rust
//!! Test inner comment
fn main() {}
```

we get this ast:
```
SOURCE_FILE@0..37
  FN@0..36
    COMMENT@0..23 "//!! Test inner comment"
    WHITESPACE@23..24 "\n"
    FN_KW@24..26 "fn"
    WHITESPACE@26..27 " "
    NAME@27..31
      IDENT@27..31 "main"
    PARAM_LIST@31..33
      L_PAREN@31..32 "("
      R_PAREN@32..33 ")"
    WHITESPACE@33..34 " "
    BLOCK_EXPR@34..36
      L_CURLY@34..35 "{"
      R_CURLY@35..36 "}"
  WHITESPACE@36..37 "\n"
```

I wonder, should we really consider inner comments as related to functions?

Reply:
This seems weird, wouldn't that inner comment belong to the containing item, so the module in this case? I suppose the parser doesn't differ between inner and outer doc comments(though imo it should)?

Reply:
Yes, counter-intuitive for me too, that's why I decided to discuss it a bit.

The https://doc.rust-lang.org/reference/comments.html does not establish any rules about that, the only thing I see:
> `//!` comments are usually used to document modules that occupy a source file.

So the way it's done now might be also fine, unless we have some other ideas to consider.

Reply:
I just checked something and interestingly enough inner attributes do apply to the parent node so this does feel like a bug that inner doc comments aren't applied to the parent. Given that `\\\` is sugar for `#[doc = ""]` and `\\!` being sugar for `#![doc =""]` they should be handled equally at least. https://doc.rust-lang.org/reference/comments.html#doc-comments

Reply:
Hm, `\\!` -> `#![doc =""]` conversion is a good catch.

So, to align the representation, does this mean we need to add a grammar, similar to https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/parser/src/grammar/attributes.rs for comments?

I've never written any before + the new syntax structure might bload the diff quite high, due to test failures it will intriduce.
So I feel like it's worth a separate PR and maybe an issue, if we don't have any.
I'll add a FIXME in this PR meanwhile.

Reply:
I'll make a fix for this shortly

Reply:
#6534